### ISSUE TITLE
Mocking utilities from another submodule

### ISSUE THREAD

**Issue by casvanluijtelaar** [lupuuss/Mokkery#40]
I'm migrating from mockative to mock. We've always had a "test-utils" submodule which also contains some mocking helpers:

```Kotlin
fun Repository.generateMocks() {
     every { foo(any()) } calls { (string: String) -> ... }
     ....
}
```
these extensions live in commonMain, so I have added:
```kotlin
mokkery {
    rule.set(ApplicationRule.Listed("commonMain"))
}
```
to the submodule's `build.gradle.kts`

However when I try to run a test using these utility method I get a

```
This call should be replaced by the Mokkery plugin! Please make sure you applied the plugin correctly!
dev.mokkery.internal.MokkeryPluginNotAppliedException: This call should be replaced by the Mokkery plugin! Please make sure you applied the plugin correctly!
	at dev.mokkery.EveryKt.every(Every.kt:19)
	at com.mcdonalds.lib.testutils.MockExtensionsKt.generateMocks(MockExtensions.kt:16)
	at com.mcdonalds.presentation.profile.ProfileViewModelTest.setup(ProfileViewModelTest.kt:68)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:112)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:40)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:60)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:52)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jd
```
If I move those mocks from a utility class to the test directly, the tests pass just fine. Probably missing something obvious

**Comment by lupuuss** [comment]
Hi! Target grouping source sets are kind of special. The rule must also include all depending source sets.
In your case, all main sources sets are needed. Change the rule to:
```kotlin
mokkery {
    rule.set { it.endsWith("Main") }
}
```

Also, maybe that's something that plugin could do automatically - pick up source sets that depend on ones selected by the rule.

**Comment by casvanluijtelaar** [comment]
unfortunately, adding

```
mokkery {
    rule.set { it.name.endsWith("Main") }
}
```
to the test-util submodule did not resolve the issue

**Comment by lupuuss** [comment]
Could you provide a project with a simple reproducer?

**Comment by casvanluijtelaar** [comment]
yes, here you go: https://github.com/casvanluijtelaar/mokkery-error-example

run the test in the "library" submodule

**Comment by lupuuss** [comment]
Is there a chance that it fails only for Android? `androidMain` splits into `androidDebug` and `androidRelease`, so the rule I provided does not include them. To be fair, in your case, you can just use `rule.set { true }`, as you need it for all source sets in `test-utils`.

**Comment by casvanluijtelaar** [comment]
iOS do work with my curretn setup!

**Comment by lupuuss** [comment]
I have a plan to refactor the rule in the next release. I should be able to provide:
* A predefined rule for main code
* Specifying source set in a way that it will apply the plugin to a subtree automatically

**Comment by lupuuss** [comment]
A quick update:
* I've added `ApplicationRule.All`
* After some research, I found that applying the plugin to a source set subtree is quite challenging. There are many cases where information about source set dependencies is unavailable during the compiler plugin setup—for example, the default hierarchy template for KMP is implicitly applied after the compiler plugins are configured. Therefore, I’ve decided to postpone this improvement until the it becomes more straightforward to implement. In the meantime, I’ve documented the plugin application to the main source code and noted that Mokkery application rules must include the dependent source sets manually.
